### PR TITLE
Find and avoid v2

### PR DESCRIPTION
**Pull request by tsampazk**
This PR adds a new **find and avoid** example. This one is much much more complicated. I will provide some general information to get you started with the review.

The algorithm used is [Maskable PPO from sb3-contrib](https://sb3-contrib.readthedocs.io/en/master/modules/ppo_mask.html). 
The logic that creates the mask for each step is contained within the method [get_action_mask](https://github.com/aidudezzz/deepworlds/blob/c320bea9ce5b35c249e8e96a59c72b9fb6f4a959/examples/find_and_avoid_v2/controllers/robot_supervisor_manager/find_and_avoid_v2_robot_supervisor.py#L330) of the environment. Proper masking is crucial for the agent to train and learn useful behaviors.

The robot used is a custom one with 13 forward-side facing distance sensor which can be set as _sonars_ or _generic_ through the code, as well as add noise, etc., 2 touch sensors in front one in the left and one in the right and of course 2 motors for differential drive.

The observation is the distance and relative angle to the target, the current motor speeds (-1.0 to 1.0), the touch sensor values, and the distance sensor values. All values are normalized appropriately. In addition the user can set a `steps window` as well as a `seconds window` to augment the observation with observations from the past, see a good description within the [docstring here](https://github.com/aidudezzz/deepworlds/blob/c320bea9ce5b35c249e8e96a59c72b9fb6f4a959/examples/find_and_avoid_v2/controllers/robot_supervisor_manager/find_and_avoid_v2_robot_supervisor.py#L466-L470).

The discrete action space consists of 5 actions to control the robot. The mapping can be found in the [docstring here](https://github.com/aidudezzz/deepworlds/blob/c320bea9ce5b35c249e8e96a59c72b9fb6f4a959/examples/find_and_avoid_v2/controllers/robot_supervisor_manager/find_and_avoid_v2_robot_supervisor.py#L910-L915).

The environment incorporates random map generation in terms of putting a collection of 25 obstacles within an arena randomly. There is the possibility to modify the difficulty and create curriculums of easy to hard maps, as is done in the default provided trainer. You can find a description on the [readme here](https://github.com/aidudezzz/deepworlds/blob/find-and-avoid-v2/examples/find_and_avoid_v2/README.md#world).

All the parameters, difficulty, etc. seen are exactly the setup i used to train the provided agent. There is huge room for experimentation in various aspects of the problem. I tried my best to test out many things and i think that the setup i reached produces a pretty good trained agent.

### CODE REVIEW

**Comment by tsampazk** [comment]
~Updating the description and also uploading a trained agent within the day.~ done

**Comment by tsampazk** [comment]
I did some git juggling to remove the merge commit from dev i did earlier and did a rebase instead. I think all went well as far as i can tell. There is some dirty history in the (310) commits that's from older versions i worked on, but i guess what's done is done now, it will get squashed eventually.

**Comment by tsampazk** [comment]
Some more git/github shenanigans happened, nothing much changed since i requested reviews, just renaming of a variable and an update of the main env docstring that had a TODO before

File: examples/find_and_avoid_v2/README.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
+# Find target & avoid obstacles v2
+
+## General
+This is the updated and expanded second version of find and avoid.
+For this example, a simple custom differential drive robot is used, equipped with
+forward-facing distance and touch sensors. The goal is to navigate to a target by modifying
+the speeds of the left and right motor, while avoiding obstacles using the distance and touch sensors.
+The agent observes its distance to the target as well as its relative facing angle, its motor speeds, the touch
+and distance sensor values and its latest action. The observation can also be augmented with observations
```
Comment (KelvinYang0320) [inline examples/find_and_avoid_v2/README.md]: Would you think the following suggestion is better?
``` suggestion
The agent observes its distance to the target, relative facing angle, motor speeds, touch
and distance sensor values, and latest action. The observation can also be augmented with observations 
```

File: examples/find_and_avoid_v2/README.md [hunk 2]
```diff
@@ -0,0 +1,51 @@
+# Find target & avoid obstacles v2
+
+## General
+This is the updated and expanded second version of find and avoid.
+For this example, a simple custom differential drive robot is used, equipped with
+forward-facing distance and touch sensors. The goal is to navigate to a target by modifying
+the speeds of the left and right motor, while avoiding obstacles using the distance and touch sensors.
```
Comment (KelvinYang0320) [inline examples/find_and_avoid_v2/README.md]:
``` suggestion
the left and right motor speeds while avoiding obstacles using the distance and touch sensors.
```

File: examples/find_and_avoid_v2/README.md [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+The agent observes its distance to the target as well as its relative facing angle, its motor speeds, the touch
+and distance sensor values and its latest action. The observation can also be augmented with observations 
+from earlier steps.
+
+## World
+The world consists of an arena that can be populated with various obstacles, boxes of various shapes,
+chairs and jars. With the help of a grid-map, there are two ways the map can be randomized. The first one
```
Comment (KelvinYang0320) [inline examples/find_and_avoid_v2/README.md]: Would you think the following suggestion is better?
``` suggestion
chairs, and jars. With the help of a grid map, the map can be randomized in two ways. The first one
```

File: examples/find_and_avoid_v2/README.md [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+places the robot and the target at random positions, with variable distance between them to control difficulty,
```
Comment (KelvinYang0320) [inline examples/find_and_avoid_v2/README.md]:
``` suggestion
places the robot and the target at random positions, with a variable distance between them to control difficulty,
```

File: examples/find_and_avoid_v2/README.md [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+and a variable number of obstacles are scattered randomly in the arena. It is made sure that there is 
+a free path on the grid map between the robot and the target using a simple BFS path-finding algorithm.
+The second way of randomizing the map, creates a corridor with several rows and three columns. The robot is
```
Comment (KelvinYang0320) [inline examples/find_and_avoid_v2/README.md]:
``` suggestion
The second way of randomizing the map creates a corridor with several rows and three columns. The robot is
```

File: examples/find_and_avoid_v2/README.md [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+placed on the first row and the target is placed on one of the other rows which can be controlled to modify difficulty.
```
Comment (KelvinYang0320) [inline examples/find_and_avoid_v2/README.md]:
``` suggestion
placed on the first row, and the target is placed on one of the other rows, which can be controlled to modify difficulty.
```

File: examples/find_and_avoid_v2/README.md [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+Two obstacles are placed in each of the in-between rows, making sure there is a path between the robot and the target.
+
+
+## Training
+The training happens in a curriculum of increasing difficulty. First a random map is used with a few scattered
```
Comment (KelvinYang0320) [inline examples/find_and_avoid_v2/README.md]:
``` suggestion
The training happens in a curriculum of increasing difficulty. First, a random map is used with a few scattered 
```

File: examples/find_and_avoid_v2/README.md [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+a free path on the grid map between the robot and the target using a simple BFS path-finding algorithm.
+The second way of randomizing the map, creates a corridor with several rows and three columns. The robot is
+placed on the first row and the target is placed on one of the other rows which can be controlled to modify difficulty.
+Two obstacles are placed in each of the in-between rows, making sure there is a path between the robot and the target.
+
+
+## Training
+The training happens in a curriculum of increasing difficulty. First a random map is used with a few scattered 
+obstacles and the target placed at a large distance. This way paths are created between the robot and the target 
+that are mostly free of obstacles, providing an easy start.  Next, using the corridor randomization, training continues
```
Comment (KelvinYang0320) [inline examples/find_and_avoid_v2/README.md]: Would you think the following suggestion is better?
``` suggestion
obstacles and the target placed at a large distance. This way, paths between the robot and the target 
are mostly free of obstacles, providing an easy start. Next, using the corridor randomization, training continues
```

File: examples/find_and_avoid_v2/README.md [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+starting with one row of obstacles and increasing gradually between training sessions, to two, three and so on, 
+with more and more obstacles added for each row.
```
Comment (KelvinYang0320) [inline examples/find_and_avoid_v2/README.md]:
``` suggestion
starting with one row of obstacles and increasing gradually between training sessions to two, three, and so on, 
with more and more obstacles added for each row. 
```

File: examples/find_and_avoid_v2/README.md [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+
+
+## Training
+The training happens in a curriculum of increasing difficulty. First a random map is used with a few scattered 
+obstacles and the target placed at a large distance. This way paths are created between the robot and the target 
+that are mostly free of obstacles, providing an easy start.  Next, using the corridor randomization, training continues
+starting with one row of obstacles and increasing gradually between training sessions, to two, three and so on, 
+with more and more obstacles added for each row. 
+Finally, the agent is trained in a random map, with all 25 available obstacles placed within and varying distance 
+between the robot and the target.
```
Comment (KelvinYang0320) [inline examples/find_and_avoid_v2/README.md]:
``` suggestion
Finally, the agent is trained on a random map, with all 25 available obstacles placed within and a varying distance 
between the robot and the target.
```

File: examples/find_and_avoid_v2/README.md [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+
+## Testing
+Testing first takes place in random maps set up similar to the last training session using stable-baselines3's 
+evaluation for 100 episodes (and consequently maps). 
+
+Then a custom evaluation procedure takes place that tests the agent in various corridor setups of increasing 
+difficulties and a random map setup, for 100 episodes each.
```
Comment (KelvinYang0320) [inline examples/find_and_avoid_v2/README.md]:
``` suggestion
difficulties and a random map setup for 100 episodes each.
```

File: examples/find_and_avoid_v2/README.md [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+
+## Logging
+Tensorboard is used to log various aspects of the training procedure. For watching the tensorboard logs, navigate to
```
Comment (KelvinYang0320) [inline examples/find_and_avoid_v2/README.md]: I think "to watch" is better.
``` suggestion
Tensorboard is used for logging various aspects of the training procedure. To watch the tensorboard logs, navigate to 
```

File: examples/find_and_avoid_v2/README.md [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+difficulties and a random map setup, for 100 episodes each.
+
+## Logging
+Tensorboard is used to log various aspects of the training procedure. For watching the tensorboard logs, navigate to 
+`/deepworlds/examples/find_and_avoid_v2/controllers/robot_supervisor_manager` and run 
+`tensorboard --logdir ./experiments/`.
+
+For testing, a CSV file is produced that includes the reward per episode, how the episode ended (collision, timeout 
+or reached target), as well as the steps taken until the end of the episode. Moreover, the sb3 evaluation mean 
+and STD reward are recorded.
```
Comment (KelvinYang0320) [inline examples/find_and_avoid_v2/README.md]: I don't quite understand "sb3 evaluation mean and STD reward".
Do you mean the mean and STD of the reward from the sb3 evaluation?
``` suggestion
For testing, a CSV file is produced that includes the reward per episode, how the episode ended (collision, timeout, 
or reached target), as well as the steps taken until the end of the episode. Moreover, the sb3 evaluated mean 
and STD reward are recorded.
```

**Review by KelvinYang0320 — CHANGES_REQUESTED** [review]
@tsampazk  I have left a few comments for the README.
Some of my suggestions may not be very good.
I will review the code another time.

**Comment by tsampazk** [comment]
Thank you @KelvinYang0320! No worries, i will apply the fixes to the readme and feel free to review the rest when you can :smile:

File: examples/find_and_avoid_v2/README.md (hunk 13 above)
Comment (tsampazk) [reply examples/find_and_avoid_v2/README.md]: Yes, the `evaluate` method returns the average reward achieved by the agent as well as the standard deviation of the reward between episodes. I will try to rephrase it in another way to make it more clear.

Comment (tsampazk) [reply examples/find_and_avoid_v2/README.md]: https://github.com/aidudezzz/deepworlds/pull/88/commits/3df70b97d52856499d1ff1c2f5c87efceae4b8f0

Let me know how this looks! :)

**Comment by tsampazk** [comment]
I'll go ahead and merge this as i need to include the regular link to the code for the relevant paper and we can revisit it later if needed. In general i feel that the example is in a great spot and only some details might need more work. As soon as the paper is published i will add it in the relevant README.

**Pull request merged**